Introduce `icicle_cpu::lifter::DecodeError` for better error handling

mrexodia opened the pull request:
Previously it was impossible to differentiate between the following errors:

- Invalid instruction (no valid decoding)
- Non-executable memory being executed
- IP alignment error

I noticed this when I set RIP to an invalid address and the exception was `InvalidInstruction`. This is my first rust code, so please review it carefully. There are also a few other errors now, but it's unclear how to handle them. Especially the `GroupStartEndEqual` is confusing and I don't know when this would happen...

mchesser commented on icicle-cpu/src/lifter/mod.rs:
```diff
@@ -678,7 +699,7 @@ impl BlockLifter {
                 }
                 BlockResult::Exit(addr) => break Target::External(addr.into()),
-                BlockResult::Invalid => break Target::Invalid,
+                BlockResult::Invalid(e) => return Err(e),
```
We can't return here directly because there might be previous instructions in the block before the invalid instruction that we still want to be able to execute them.

The strategy I was using here was to was basically to allow blocks that end in an invalid instruction to still be generated then handle actually raising it to an exception when the end of the block is reached.

mchesser replied:
I think what should be done here is that the decode error should be injected as `pcode::Op::Exception` to allow things to work better.

(Ideally also handling the optimization when there are not extra instructions in the block (like below))

mrexodia replied:
Thanks for the feedback! This makes sense and I'll create a test case for it and see if I can get it to work. This also explains my confusion with the `BlockStartEndEqual`, this would have to be replaced with the appropriate `DecodeError`.

mchesser commented on icicle-cpu/src/lifter/mod.rs:
```diff
@@ -702,12 +723,13 @@ impl BlockLifter {
             }
         }
 
+        // TODO: When does this error happen exactly?
```
This occurs when the block starts with an invalid instruction there is no need to generate a block at all, and we can immediately raise an exception.

mchesser reviewed (commented):
Thanks for this.

It would be good to have something like this, however it think it will prevent blocks that start with valid instructions from being executed (see the comments above).

mchesser commented on icicle-vm/src/lib.rs:
```diff
@@ -291,13 +291,20 @@ impl Vm {
-                self.cpu.exception = cpu::Exception::new(ExceptionCode::InvalidInstruction, pc);
+            Err(e) => {
+                tracing::trace!("DecodeError at {pc:X}: {e:?}");
+                let code = match e {
```
It's probably worth implementing the `From` trait (e.g.: https://github.com/icicle-emu/icicle-emu/blob/2706b71b0275b0ceae2aa0ee9b83aef1866213b4/icicle-cpu/src/lib.rs#L312) to match how the rest of individual subsystem errors are mapped.

mrexodia commented:
Ping @mchesser I recreated the PR, now everything looks like it works correctly!

mrexodia commented:
This also adds a new exception code ExecUnaligned, not sure if that needs any additional handling…

rbran commented on icicle-cpu/src/lifter/mod.rs:
```diff
@@ -73,17 +82,18 @@ impl InstructionLifter {
 
-        let block = self.lifter.lift(&src.arch().sleigh, &self.decoded).ok()?;
+        let block = self.lifter.lift(&src.arch().sleigh, &self.decoded)
+            .map_err(|_| DecodeError::InvalidInstruction)?;
```
BTW, If the lifter fails because it's unable to attach a token to a varnode (`UnknownVarNode`) should the disassembler backtrack and try the next constructor?

This is probably too much of an edge case that never happens, but having those errors separated could be helpful to find-out in the future.

mchesser commented on icicle-cpu/src/lifter/mod.rs:
```diff
@@ -997,12 +1029,12 @@ impl BlockExit {
     fn fmt(&self, f: &mut std::fmt::Formatter, ctx: &T) -> std::fmt::Result {
         match self {
-            Self::Jump { target: Target::Invalid } => f.write_str("invalid_instruction"),
+            Self::Jump { target: Target::Invalid(_, _) } => f.write_str("invalid_instruction"), // TODO: handle error specifically?
```
Similar to above -- this is used for emulator debugging, so extra info could be could to print.

mchesser commented on icicle-vm/src/lib.rs:
```diff
@@ -291,13 +291,14 @@ impl Vm {
-            None => {
-                self.cpu.exception = cpu::Exception::new(ExceptionCode::InvalidInstruction, pc);
+            Err(e) => {
+                tracing::trace!("DecodeError at {pc:X}: {e:?}");
```
```suggestion
                tracing::trace!("DecodeError at {pc:#x}: {e:?}");
```

(very minor -- everywhere else, the code uses the lowercase hex formatting (i.e. `:#x) so could be changed for consistency)

(Also just fyi, (if you were not already aware), the `#` out the front causes the `0x` prefix to be printed)

mchesser reviewed (commented):
This looks pretty good now (close to how I would have implemented it). Thanks!

There are a few minor issues (most should be in comments above). In addition there are couple of other changes needed due to API changes.

https://github.com/icicle-emu/icicle-emu/blob/cea57f5a57ed8271bca08159a22e3edbebd74a0a/icicle-test/src/main.rs#L189

Should be changed to:

```rust
                .map_err(|e| anyhow::format_err!("Failed to lift: {:#0x}: {e:?}", test.load_addr))?;
```
And https://github.com/icicle-emu/icicle-emu/blob/cea57f5a57ed8271bca08159a22e3edbebd74a0a/icicle-test/src/tester.rs#L49 and https://github.com/icicle-emu/icicle-emu/blob/cea57f5a57ed8271bca08159a22e3edbebd74a0a/icicle-test/src/tester.rs#L101

Need to change to `if let Ok(group) = group {` 


I also noticed there are some miscellaneous formatting changes left behind (which I think is probably because of my custom `rustfmt` config see https://github.com/icicle-emu/icicle-emu/pull/39#issuecomment-1607408301).

If you run `cargo +nightly fmt` (from the project root) it should clean them up (as long as you have nightly installed) -- otherwise I can just fix the formatting up after the merge.


For the TODOs, feel free to address them or leave them in, I have some additional changes I'm planning on merging in soon, and I can address them then when

mrexodia commented:
I addressed the comments and suggestions that were clearly actionable for me. I tried running the format, but this changes a lot of files unrelated to the ones I changed too, so I think it's better to merge and then for you to format as you prefer. For the `DecodeError:: InvalidLiftCode` I'm not sure what to do so probably best if you address this as you see fit, I don't have a deep enough understanding of the code...

mchesser commented:
Thanks I've merged it now.
